File: src/main/java/org/cobbzilla/util/io/FilesystemWalker.java
package org.cobbzilla.util.io;

import lombok.experimental.Accessors;
import org.cobbzilla.util.daemon.AwaitResult;
import org.cobbzilla.util.string.StringUtil;
import org.slf4j.Logger;

import java.io.File;
import java.io.FileFilter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import static org.cobbzilla.util.daemon.Await.awaitAll;
import static org.cobbzilla.util.daemon.DaemonThreadFactory.fixedPool;
import static org.cobbzilla.util.io.FileUtil.abs;
import static org.cobbzilla.util.io.FileUtil.isSymlink;
import static org.cobbzilla.util.system.Sleep.sleep;
import static org.cobbzilla.util.time.TimeUtil.parseDuration;

@Accessors(chain=true)
public class FilesystemWalker {

    private static final Logger log = org.slf4j.LoggerFactory.getLogger(FilesystemWalker.class);
    private final List<File> dirs = new ArrayList<>();
    private final List<FilesystemVisitor> visitors = new ArrayList<>();
    private boolean includeSymlinks = true;
    private boolean visitDirs = false;
    private int threads = 5;
    private int size = 1_000_000;
    private long timeout = TimeUnit.MINUTES.toMillis(15);
    private FileFilter filter;
    private long sleepTime = TimeUnit.SECONDS.toMillis(5);

    public boolean hasFilter () { return filter != null; }

    public FilesystemWalker withDir (File dir) { dirs.add(dir); return this; }
    public FilesystemWalker withDirs (List<File> dirs) { this.dirs.addAll(dirs); return this; }
    public FilesystemWalker withDirs (File[] dirs) { this.dirs.addAll(Arrays.asList(dirs)); return this; }
    public FilesystemWalker withVisitor (FilesystemVisitor visitor) { visitors.add(visitor); return this; }
    public FilesystemWalker withTimeoutDuration (String duration) { setTimeout(parseDuration(duration)); return this; }

    private final ExecutorService pool = fixedPool(getThreads());
    private final List<Future<?>> futures = new ArrayList<>(getSize());

    public AwaitResult walk() {
        for (File dir : dirs) fileJob(dir);

        // wait for number of futures to stop increasing
        do {
            final int lastNumFutures = numFutures();
            awaitFutures();
            if (numFutures() == lastNumFutures) break;
            sleep(getSleepTime());
        } while (true);
        return awaitFutures();
    }

    private AwaitResult awaitFutures() {
        final AwaitResult result = awaitAll(getFutures(), getTimeout());
        if (!result.allSucceeded()) log.warn(StringUtil.toString(result.getFailures().values(), "\n---------"));
        return result;
    }

    private int numFutures () {
        final List<Future<?>> futures = getFutures();
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized (futures) { return futures.size(); }
    }

    private boolean fileJob(File f) {
        final List<Future<?>> futures = getFutures();
        final Future<?> future = getPool().submit(new FsWalker(f));
        //noinspection SynchronizationOnLocalVariableOrMethodParameter
        synchronized (futures) { return futures.add(future); }
    }

    public List<File> getDirs() {
        return this.dirs;
    }

    public List<FilesystemVisitor> getVisitors() {
        return this.visitors;
    }

    public boolean isIncludeSymlinks() {
        return this.includeSymlinks;
    }

    public boolean isVisitDirs() {
        return this.visitDirs;
    }

    public int getThreads() {
        return this.threads;
    }

    public int getSize() {
        return this.size;
    }

    public long getTimeout() {
        return this.timeout;
    }

    public FileFilter getFilter() {
        return this.filter;
    }

    public long getSleepTime() {
        return this.sleepTime;
    }

    public ExecutorService getPool() {
        return this.pool;
    }

    public List<Future<?>> getFutures() {
        return this.futures;
    }

    public FilesystemWalker setIncludeSymlinks(boolean includeSymlinks) {
        this.includeSymlinks = includeSymlinks;
        return this;
    }

    public FilesystemWalker setVisitDirs(boolean visitDirs) {
        this.visitDirs = visitDirs;
        return this;
    }

    public FilesystemWalker setThreads(int threads) {
        this.threads = threads;
        return this;
    }

    public FilesystemWalker setSize(int size) {
        this.size = size;
        return this;
    }

    public FilesystemWalker setTimeout(long timeout) {
        this.timeout = timeout;
        return this;
    }

    public FilesystemWalker setFilter(FileFilter filter) {
        this.filter = filter;
        return this;
    }

    public FilesystemWalker setSleepTime(long sleepTime) {
        this.sleepTime = sleepTime;
        return this;
    }

    private class FsWalker implements Runnable {
        private final File file;

        @java.beans.ConstructorProperties({"file"})
        public FsWalker(File file) {
            this.file = file;
        }

        @Override public void run() {
            if (isSymlink(file) && !includeSymlinks) return;

            if (file.isFile()) {
                // visit the file
                visit();

            } else if (file.isDirectory()) {
                // should we visit directory entries?
                if (visitDirs) visit();

                // should we filter directory entries?
                final File[] files = hasFilter() ? file.listFiles(filter) : file.listFiles();

                // walk each entry in the directory
                if (files != null) for (File f : files) fileJob(f);

            } else {
                log.warn("unexpected file: neither file nor directory, skipping: "+abs(file));
            }

        }

        private void visit() { for (FilesystemVisitor visitor : getVisitors()) visitor.visit(file); }
    }
}

File: src/main/java/org/cobbzilla/util/io/FilesystemWalker.kt
package org.cobbzilla.util.io

import lombok.experimental.Accessors
import org.cobbzilla.util.daemon.AwaitResult
import org.cobbzilla.util.string.StringUtil
import org.slf4j.Logger

import java.io.File
import java.io.FileFilter
import java.util.ArrayList
import java.util.Arrays
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit

import org.cobbzilla.util.daemon.Await.awaitAll
import org.cobbzilla.util.daemon.DaemonThreadFactory.fixedPool
import org.cobbzilla.util.io.FileUtil.abs
import org.cobbzilla.util.io.FileUtil.isSymlink
import org.cobbzilla.util.system.Sleep.sleep
import org.cobbzilla.util.time.TimeUtil.parseDuration

@Accessors(chain = true)
class FilesystemWalker {
    private val dirs = ArrayList<File>()
    private val visitors = ArrayList<FilesystemVisitor>()
    private var includeSymlinks = true
    private var visitDirs = false
    private var threads = 5
    private var size = 1000000
    private var timeout = TimeUnit.MINUTES.toMillis(15)
    private var filter: FileFilter? = null
    private var sleepTime = TimeUnit.SECONDS.toMillis(5)

    val pool = fixedPool(getThreads())
    private val futures = ArrayList<Future<*>>(getSize())

    fun hasFilter(): Boolean {
        return filter != null
    }

    fun withDir(dir: File): FilesystemWalker {
        dirs.add(dir)
        return this
    }

    fun withDirs(dirs: List<File>): FilesystemWalker {
        this.dirs.addAll(dirs)
        return this
    }

    fun withDirs(dirs: Array<File>): FilesystemWalker {
        this.dirs.addAll(Arrays.asList(*dirs))
        return this
    }

    fun withVisitor(visitor: FilesystemVisitor): FilesystemWalker {
        visitors.add(visitor)
        return this
    }

    fun withTimeoutDuration(duration: String): FilesystemWalker {
        setTimeout(parseDuration(duration))
        return this
    }

    fun walk(): AwaitResult<*> {
        for (dir in dirs) fileJob(dir)

        // wait for number of futures to stop increasing
        do {
            val lastNumFutures = numFutures()
            awaitFutures()
            if (numFutures() == lastNumFutures) break
            sleep(getSleepTime())
        } while (true)
        return awaitFutures()
    }

    private fun awaitFutures(): AwaitResult<*> {
        val result = awaitAll<Any>(getFutures(), getTimeout())
        if (!result.allSucceeded()) log.warn(StringUtil.toString(result.failures.values, "\n---------"))
        return result
    }

    private fun numFutures(): Int {
        val futures = getFutures()

        synchronized(futures) {
            return futures.size
        }
    }

    private fun fileJob(f: File): Boolean {
        val futures = getFutures()
        val future = pool.submit(FsWalker(f))

        synchronized(futures) {
            return futures.add(future)
        }
    }

    fun getDirs(): List<File> {
        return this.dirs
    }

    fun getVisitors(): List<FilesystemVisitor> {
        return this.visitors
    }

    fun isIncludeSymlinks(): Boolean {
        return this.includeSymlinks
    }

    fun isVisitDirs(): Boolean {
        return this.visitDirs
    }

    fun getThreads(): Int {
        return this.threads
    }

    fun getSize(): Int {
        return this.size
    }

    fun getTimeout(): Long {
        return this.timeout
    }

    fun getFilter(): FileFilter? {
        return this.filter
    }

    fun getSleepTime(): Long {
        return this.sleepTime
    }

    fun getFutures(): MutableList<Future<*>> {
        return this.futures
    }

    fun setIncludeSymlinks(includeSymlinks: Boolean): FilesystemWalker {
        this.includeSymlinks = includeSymlinks
        return this
    }

    fun setVisitDirs(visitDirs: Boolean): FilesystemWalker {
        this.visitDirs = visitDirs
        return this
    }

    fun setThreads(threads: Int): FilesystemWalker {
        this.threads = threads
        return this
    }

    fun setSize(size: Int): FilesystemWalker {
        this.size = size
        return this
    }

    fun setTimeout(timeout: Long): FilesystemWalker {
        this.timeout = timeout
        return this
    }

    fun setFilter(filter: FileFilter): FilesystemWalker {
        this.filter = filter
        return this
    }

    fun setSleepTime(sleepTime: Long): FilesystemWalker {
        this.sleepTime = sleepTime
        return this
    }

    private inner class FsWalker @java.beans.ConstructorProperties("file")
    constructor(private val file: File) : Runnable {

        override fun run() {
            if (isSymlink(file) && !includeSymlinks) return

            if (file.isFile) {
                // visit the file
                visit()

            } else if (file.isDirectory) {
                // should we visit directory entries?
                if (visitDirs) visit()

                // should we filter directory entries?
                val files = if (hasFilter()) file.listFiles(filter) else file.listFiles()

                // walk each entry in the directory
                if (files != null) for (f in files) fileJob(f)

            } else {
                log.warn("unexpected file: neither file nor directory, skipping: " + abs(file))
            }

        }

        private fun visit() {
            for (visitor in getVisitors()) visitor.visit(file)
        }
    }

    companion object {

        private val log = org.slf4j.LoggerFactory.getLogger(FilesystemWalker::class.java)
    }
}
